artemis-server/src/main/java/org/apache/activemq/artemis/core/remoting/server/impl/RemotingServiceImpl.java

```diff
@@ -407,15 +410,28 @@ public void stop(final boolean criticalError) throws Exception {
          conn.disconnect(criticalError);
       }
 
+      Map<Acceptor, Future<?>> acceptorFutures = new HashMap<>();
       for (Acceptor acceptor : acceptors.values()) {
          try {
-            acceptor.stop();
+            //In some cases an acceptor stopping could be locked ie NettyAcceptor stopping could be locked by a network failure.
+            acceptorFutures.put(acceptor, acceptor.asyncStop());
```

Comment:
it would be a lot simpler if instead of returning a future, you passed in a Runnable or Future

You could have the following code:

```java
// Instead of a Map<Acceptor, Future>
CountDownLatch latch = new CountDownLatch(acceptors.size());

for (Acceptor accetor: acceptors.values()) {
    acceptor.asyncStop(latch::countDown);
}
```

and the whole loop on each Future would be replaced by:
```java
   latch.await();
```